### PR TITLE
[MRG] CSD plot smoothing and scaling

### LINKED ISSUE #68: add CSD plot for comparing against animal studies
### Background

The ability to calculate LFP for simulated electrodes in the model (#150) allow for time vs. depth laminar CSD plots to be generated. Laminar CSD (electrodes arranged vertically) is the primary goal of this feature because laminar CSD is commonly used in animal studies where a laminar probe is inserted into the cortext to measure extracellular potentials. In the experimental animal studies, CSD is calculated from the bandpass filtered LFP data (low-frequency portion of extracellular potentials) to identify underlying neuronal activity at specific cortical layers. Such activity is identified as a "sink" (net influx of current into cell) or "source" (efflux into extracellular space) in the temporal-spatial domain.

Figure 5 from Sherman et al. 2016 shows laminar CSD plots of an anesthetized mouse and an awake monkey.

Screen Shot 2020-09-08 at 8 41 05 AM

> Sherman MA, Lee S, Law R, et al. Neural mechanisms of transient neocortical beta rhythms: Converging evidence from humans, computational modeling, monkeys, and mice. Proc Natl Acad Sci U S A. 2016;113(33):E4885-E4894. doi:10.1073/pnas.1604135113

CSD can be calculated via
```
numpy.diff(LFP_data,n=2,axis=ax)/spacing_mm**2
```

The sign may be adjusted to make a current sinks negative and sources positive.

### Goal

Create laminar CSD plots from simulated data that are comparable to Figure 5 in Sherman et al. 2016

### Requirements

1. Averaging LFP over multiple trials is typically done before computing CSD
2. CSD plots are created using a 2D spline over values at each electrode depth. Support various spline functions.
3. Users need to be able to easily adjust the X,Y scales to match simulated data with their experimental data
4. The CSD of the first and last channels cannot be computed using the 2nd spatial derivative. The Vaknin method is used to derive approximate CSD signals at these points

    > G. Vaknin, P.G. DiScenna, T.J. Teyler, A method for calculating current source density (CSD) analysis without resorting to recording sites outside the sampling volume, Journal of Neuroscience Methods, Volume 24, Issue 2, 1988, Pages 131-135, ISSN 0165-0270, https://doi.org/10.1016/0165-0270(88)90056-8.
5. Implement [iCSD method](https://github.com/espenhgn/iCSD) as an alternative to `numpy.diff()` above

### PR DESCRIPTION
**Pull request by katduecker**
This PR closes #68.
I have added an option to smoothen the CSD plot using spline interpolation, to set the range of the colormap using vmin and vmax, and to set the color of the sink (and source).

If vmin and vmax are None, the colormap is set automatically based on the maximum absolute value of CSD, such that green is zero.

### CODE REVIEW

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1254]:
```suggestion
    interpolation : str | None
```

File: hnn_core/viz.py
Comment (ntolley) [inline hnn_core/viz.py:1249]: Can you have vmin/vmax as two separate entries?

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1251]:
```suggestion
    sink : str
```

There is a very specific formatting of this docstring ... you should describe the argument below and only mention the datatype here

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1275]: what if the user passes 'B' ? I have seen people doing things like that ...

Comment (jasmainak) [reply hnn_core/viz.py:1275]: you can do something like:

```py
if sink[0].tolower() == 'b'
```

but in addition, also throw an error if the string is something other than 'r' or 'b'

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1295]: wondering if we should use `imshow`? I know this is how it was implemented originally. The advantage of `imshow` is that you can mouseover to see the value.

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1285]:
```suggestion
    elif interpolation is None:
```

in python `None` is a specific data type. It's not a `boolean` that can be True or False. Better to be explicit, otherwise the user may pass `interpolation=False` and that may also work but it is unintentional

File: doc/whats_new.rst
Comment (jasmainak) [inline doc/whats_new.rst:79]:
```suggestion
  by `Katharina Duecker`_ in :gh:`815`
```

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1270]: why did you remove these?

Comment (katduecker) [reply hnn_core/viz.py:1270]: because the labels are not set in other plotting functions

File: hnn_core/viz.py
Comment (katduecker) [reply hnn_core/viz.py:1295]: I tried this and imshow gets messy when interpolation = None. The aspect ratio of the axes has to be set, the origin has to be flipped, and the ticklabels for x and y have to be set. Maybe more complicated that necessary for this PR?

Comment (jasmainak) [reply hnn_core/viz.py:1302]: oh dear! agreed

File: examples/howto/plot_record_extracellular_potentials.py
Comment (jasmainak) [inline examples/howto/plot_record_extracellular_potentials.py:104]: perhaps we leave this out? Or break this into two lines? Looks a bit intimidating to a new python user right now

File: hnn_core/tests/test_extracellular.py
Comment (jasmainak) [inline hnn_core/tests/test_extracellular.py:285]:
```suggestion
    with pytest.raises(RuntimeError, match='Please use sink = "b" or '
```

python trick ... you just need to use double quotes inside single quotes or single quotes inside double quotes and it will work

File: hnn_core/viz.py
Comment (jasmainak) [inline hnn_core/viz.py:1281]: same here!

**Review by jasmainak — APPROVED** [review]
Looks clean otherwise. Merge button belongs to @ntolley and @gtdang

File: examples/howto/plot_record_extracellular_potentials.py
Comment (ntolley) [reply examples/howto/plot_record_extracellular_potentials.py:104]: I think this was originally my suggestion just to make users aware of these options

point is well taken though, it might be better to just show a short function that makes nice looking plot with no fuss, I'd be fine to see it removed

File: hnn_core/viz.py
Comment (ntolley) [reply hnn_core/viz.py:1270]: I'll vote keeping in this case, maybe we should actually consider adding these to the default plots...

File: hnn_core/tests/test_extracellular.py
Comment (katduecker) [reply hnn_core/tests/test_extracellular.py:285]: ah, good to know I was breaking my ears on this one :)

**Comment by katduecker** [comment]
Should be okay now @ntolley @gtdang

**Comment by katduecker** [comment]
@ntolley @gtdang @jasmainak I suspect that I'm tracking some files that I shouldn't be tracking.. (I used some tracking feature in VSCode which I probably shouldn't have..) can you give me a pointer to what you think is going on?

**Comment by katduecker** [comment]
@ntolley @gtdang is the ubuntu check due to something I have done?

File: examples/howto/plot_record_extracellular_potentials.py
Comment (ntolley) [inline examples/howto/plot_record_extracellular_potentials.py:101]: Was this left over from when these options were exposed in the plot call?

Or did you intentionally add it _because_ we decided to remove these options for simplicity?

Your call but I just wanted to point it out

**Review by ntolley — APPROVED** [review]
Looks great on my end @katduecker!!

Small comment to address and then I'll go ahead and merge

Also that test (link check) fails occasionally if a website's URL linked through our documentation is down, usually just need to rerun the job and it'll pass

File: examples/howto/plot_record_extracellular_potentials.py
Comment (katduecker) [reply examples/howto/plot_record_extracellular_potentials.py]: Whoops no, that was supposed to go. Changed it!

**Pull request merged**
